### ISSUE TITLE
Imagine can't work with PNG-24 image on ImageMagick 7

### ISSUE THREAD

**Issue by ossinkine** [avalanche123/Imagine#689]
### Issue description

I've updated ImageImagick version 6.9.5 to 7.0.7 and image thumbnails in my app became empty image.
I spent some time to debug and I've localized an issue.
Removing [this line](https://github.com/avalanche123/Imagine/blob/v0.7.1/lib/Imagine/Imagick/Image.php#L841) fixes the issue. Replacing [the constant](https://github.com/avalanche123/Imagine/blob/v0.7.1/lib/Imagine/Imagick/Image.php#L833) `IMGTYPE_TRUECOLORMATTE` with `IMGTYPE_TRUECOLOR` also fixes the issue.
I think this issue relates to #524.

### What's the PHP version you are using?

`7.1.23`

### What's the imaging library you are using [gd/imagick/gmagick/any]?

`imagick`

### What's the imaging library configuration 

```
imagick module => enabled
imagick module version => 3.4.3
imagick classes => Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version => ImageMagick 7.0.7-32 Q16 x86_64 2018-05-15 https://www.imagemagick.org
Imagick using ImageMagick library version => ImageMagick 7.0.7-32 Q16 x86_64 2018-05-15 https://www.imagemagick.org
ImageMagick copyright => © 1999-2018 ImageMagick Studio LLC
ImageMagick release date => 2018-05-15
ImageMagick number of supported formats:  => 224
ImageMagick supported formats => 3FR, 3G2, 3GP, AAI, AI, ART, ARW, AVI, AVS, BGR, BGRA, BGRO, BMP, BMP2, BMP3, BRF, CAL, CALS, CANVAS, CAPTION, CIN, CIP, CLIP, CMYK, CMYKA, CR2, CRW, CUR, CUT, DCM, DCR, DCX, DDS, DFONT, DNG, DOT, DPX, DXT1, DXT5, EPDF, EPI, EPS, EPS2, EPS3, EPSF, EPSI, EPT, EPT2, EPT3, ERF, FAX, FILE, FITS, FRACTAL, FTP, FTS, G3, G4, GIF, GIF87, GRADIENT, GRAY, GRAYA, GROUP4, GV, HALD, HDR, HISTOGRAM, HRZ, HTM, HTML, HTTP, HTTPS, ICB, ICO, ICON, IIQ, INFO, INLINE, IPL, ISOBRL, ISOBRL6, JNG, JNX, JPE, JPEG, JPG, JPS, JSON, K25, KDC, LABEL, M2V, M4V, MAC, MAP, MASK, MAT, MATTE, MEF, MIFF, MKV, MNG, MONO, MOV, MP4, MPC, MPEG, MPG, MRW, MSL, MSVG, MTV, MVG, NEF, NRW, NULL, ORF, OTB, OTF, PAL, PALM, PAM, PANGO, PATTERN, PBM, PCD, PCDS, PCL, PCT, PCX, PDB, PDF, PDFA, PEF, PES, PFA, PFB, PFM, PGM, PGX, PICON, PICT, PIX, PJPEG, PLASMA, PNG, PNG00, PNG24, PNG32, PNG48, PNG64, PNG8, PNM, PPM, PS, PS2, PS3, PSB, PSD, PTIF, PWP, RADIAL-GRADIENT, RAF, RAS, RAW, RGB, RGBA, RGBO, RGF, RLA, RLE, RMF, RW2, SCR, SCT, SFW, SGI, SHTML, SIX, SIXEL, SPARSE-COLOR, SR2, SRF, STEGANO, SUN, SVG, SVGZ, TEXT, TGA, THUMBNAIL, TIFF, TIFF64, TILE, TIM, TTC, TTF, TXT, UBRL, UBRL6, UIL, UYVY, VDA, VICAR, VID, VIFF, VIPS, VST, WBMP, WEBP, WMV, WPG, X, X3F, XBM, XC, XCF, XPM, XPS, XV, XWD, YCbCr, YCbCrA, YUV

Directive => Local Value => Master Value
imagick.locale_fix => 0 => 0
imagick.skip_version_check => 0 => 0
imagick.progress_monitor => 0 => 0
```

### Minimal PHP code to reproduce the error: 

```php

$imagine = new Imagine\Imagick\Imagine();
$image = $imagine->open($source);
$image->save($destination);

```
`$source` is the path for PNG-24 image.
`$destination` is the path where will be an empty image after running this script.

**Comment by mlocati** [comment]
@ossinkine do you have a sample image?

**Comment by ossinkine** [comment]
![slackhq](https://user-images.githubusercontent.com/2588533/47294617-1178d400-d627-11e8-8b4b-1e52e7093c44.png)

**Comment by mlocati** [comment]
@ossinkine Imagine is working fine for me with ImageMagick 7.

Here's my configuration:

```sh
$ php --ri imagick

imagick

imagick module => enabled
imagick module version => 3.4.3
imagick classes => Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version => ImageMagick 7.0.8-13 Q16 x86_64 2018-10-22 https://imagemagick.org
Imagick using ImageMagick library version => ImageMagick 7.0.8-13 Q16 x86_64 2018-10-22 https://imagemagick.org
ImageMagick copyright => © 1999-2018 ImageMagick Studio LLC
ImageMagick release date => 2018-10-22
ImageMagick number of supported formats:  => 238
ImageMagick supported formats => 3FR, 3G2, 3GP, A, AAI, AI, ART, ARW, AVI, AVS, B, BGR, BGRA, BGRO, BIE, BMP, BMP2, BMP3, BRF, C, CAL, CALS, CANVAS, CAPTION, CIN, CIP, CLIP, CMYK, CMYKA, CR2, CRW, CUR, CUT, DCM, DCR, DCRAW, DCX, DDS, DFONT, DJVU, DNG, DPX, DXT1, DXT5, EPDF, EPI, EPS, EPS2, EPS3, EPSF, EPSI, EPT, EPT2, EPT3, ERF, FAX, FILE, FITS, FRACTAL, FTP, FTS, G, G3, G4, GIF, GIF87, GRADIENT, GRAY, GRAYA, GROUP4, HALD, HDR, HISTOGRAM, HRZ, HTM, HTML, HTTP, HTTPS, ICB, ICO, ICON, IIQ, INFO, INLINE, IPL, ISOBRL, ISOBRL6, JBG, JBIG, JNG, JNX, JPE, JPEG, JPG, JPS, JSON, K, K25, KDC, LABEL, M, M2V, M4V, MAC, MAP, MASK, MAT, MATTE, MEF, MIFF, MKV, MNG, MONO, MOV, MP4, MPC, MPEG, MPG, MRW, MSL, MSVG, MTV, MVG, NEF, NRW, NULL, O, ORF, OTB, OTF, PAL, PALM, PAM, PANGO, PATTERN, PBM, PCD, PCDS, PCL, PCT, PCX, PDB, PDF, PDFA, PEF, PES, PFA, PFB, PFM, PGM, PGX, PICON, PICT, PIX, PJPEG, PLASMA, PNG, PNG00, PNG24, PNG32, PNG48, PNG64, PNG8, PNM, PPM, PS, PS2, PS3, PSB, PSD, PTIF, PWP, R, RADIAL-GRADIENT, RAF, RAS, RAW, RGB, RGBA, RGBO, RGF, RLA, RLE, RMF, RW2, SCR, SCREENSHOT, SCT, SFW, SGI, SHTML, SIX, SIXEL, SPARSE-COLOR, SR2, SRF, STEGANO, SUN, SVG, SVGZ, TEXT, TGA, THUMBNAIL, TIFF, TIFF64, TILE, TIM, TTC, TTF, TXT, UBRL, UBRL6, UIL, UYVY, VDA, VICAR, VID, VIFF, VIPS, VST, WBMP, WMF, WMV, WMZ, WPG, X, X3F, XBM, XC, XCF, XPM, XPS, XV, XWD, Y, YCbCr, YCbCrA, YUV

Directive => Local Value => Master Value
imagick.locale_fix => 0 => 0
imagick.skip_version_check => 0 => 0
imagick.progress_monitor => 0 => 0
```

Sample code (where `slackhq.png` is [the file you provided](https://github.com/avalanche123/Imagine/issues/689#issuecomment-431833522)):

```php
<?php
require_once './vendor/autoload.php';
$imagine = new Imagine\Imagick\Imagine();
$image = $imagine->open('./slackhq.png');
$image->save('./slackhq.out.png');
```

Which saves this file:
![slackhq out](https://user-images.githubusercontent.com/928116/47297166-a32b1600-d614-11e8-9773-bb2421b417e0.png)

PS: be sure to use the most recent Imagine release.

**Comment by mlocati** [comment]
Could you try to add this line right after https://github.com/avalanche123/Imagine/blob/1.0.1/src/Imagick/Image.php#L968 helps?

```php
$this->imagick->setImageType($typeMapping[$palette->name()]);
```

**Comment by ossinkine** [comment]
@mlocati This works for me

**Comment by ossinkine** [comment]
@mlocati Should I create the PR?

**Comment by mlocati** [comment]
It's already there: https://github.com/avalanche123/Imagine/pull/691

BTW I don't think that's correct: it seems that it mangles images with non-RGB color spaces.

I'm trying to see if it's a problem of ImageMagick 7.0.7-32 (from the [release notes](https://imagemagick.org/script/changelog.php) of 7.0.7-33 we read `Restore SetImageAlpha() behavior.`).

**Comment by mlocati** [comment]
Yes, I can confirm it's a problem of ImageMagick version 7.0.7-32.

Here's the result of my tests:
- 7.0.8-13: works
- 7.0.7-39: works
- 7.0.7-33: works
- 7.0.7-32: **fails**
- 7.0.7-31: works

So I guess you have to manually build ImageMagick instead of relying of the one provided by apk (see the [`.travis/imagick.sh`](https://github.com/avalanche123/Imagine/blob/develop/.travis/imagick.sh) file to see an example of how to do it)

**Comment by mlocati** [comment]
> Is it possible to check the ImageMagic version in Imagine or using Composer?

A foreword:
- ImageMagick is an application/library that's installed at the system level
- imagick is the PHP extension that uses ImageMagick to do its stuff

Detecting the ImageMagick version used by imagick could be done by inspecting the output of this command:

```sh
php --ri imagick
```

for instance with:

```sh
php --ri imagick \
  | grep -E 'Imagick using.* ImageMagick ([0-9][^ ]+)' \
  | sed -e 's/^.* ImageMagick //' \
  | sed -e 's/ .*//'
```

BTW this can only be done once ImageMagick and imagick have been installed: you may want to do this check *before* installing ImageMagick/imagick.

You can do that for instance with:
```sh
apk add --no-cache --simulate imagemagick \
  | grep -i 'Installing imagemagick ' \
  | sed -e 's/^.*imagemagick (//' \
  | sed -e 's/-r[0-9]*)$//'
```

**Comment by mlocati** [comment]
In your setup script you could for example do something like this:

```sh
IMAGEMAGICK_VERSION=`apk add --no-cache --simulate imagemagick | grep -i 'Installing imagemagick ' | sed -e 's/^.*imagemagick (//' | sed -e 's/-r[0-9]*)$//'`
if [ "$IMAGEMAGICK_VERSION" = '7.0.7.32' ]; then
	echo 'Install imagemagick and imagick from source'
else
	echo 'Install imagemagick and imagick using apk'
fi
```

**Comment by ossinkine** [comment]
@mlocati I mean would be good if Imagine can detect by itself can it work or not.
For example, I've installed Imagine in my project and try to process images. The image proccessing is failing and I have no idea what is wrong (Exactly like in my case).
But if Imagine throws exception (or just warning) it would save me a lot of time.
Something like [this](https://github.com/avalanche123/Imagine/blob/1.0.1/src/Imagick/Image.php#L709).
Or adding a workaround for broken version would be good solution.

**Comment by mlocati** [comment]
Good point!

**Comment by mlocati** [comment]
See #692

**Comment by ossinkine** [comment]
@mlocati Looks good for me 👍

**Comment by mlocati** [comment]
PS: I just released version 1.0.2, which includes #692

**Comment by bscheshirwork** [comment]
alpine3.8 contain wersion 7.0.7
https://pkgs.alpinelinux.org/packages?name=imagemagick-dev&branch=v3.8
I will have to use edge
https://pkgs.alpinelinux.org/packages?name=imagemagick-dev&branch=edge
```dockerfile
# imagick
    printf "\n%s\n%s" "@edge http://dl-cdn.alpinelinux.org/alpine/edge/main" "@testing http://dl-cdn.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories && \
    apk --update add imagemagick-dev@edge imagemagick@edge && \
    apk add --no-cache --virtual .phpize-deps $PHPIZE_DEPS libtool && \
    export CFLAGS="$PHP_CFLAGS" CPPFLAGS="$PHP_CPPFLAGS" LDFLAGS="$PHP_LDFLAGS" && \
    pecl install imagick-3.4.3 && \
    docker-php-ext-enable imagick && \
...
```
